File: 981/981.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/981.py

import java.math.BigInteger

val MODVAL: Long = 888888883L

fun qbinomMinus1Int(n: Int, k: Int): BigInteger {
    if ((n and 1) == 0 && (k and 1) == 1) return BigInteger.ZERO
    return bigComb(n shr 1, k shr 1)
}

fun bigComb(n: Int, k: Int): BigInteger {
    if (k < 0 || k > n) return BigInteger.ZERO
    if (k == 0 || k == n) return BigInteger.ONE
    var num = BigInteger.ONE
    var den = BigInteger.ONE
    val kk = if (k < n - k) k else n - k
    for (i in 0 until kk) {
        num = num * BigInteger.valueOf((n - i).toLong())
        den = den * BigInteger.valueOf((i + 1).toLong())
    }
    return num / den
}

fun bigFactorial(n: Int): BigInteger {
    var result = BigInteger.ONE
    for (i in 2..n) result = result * BigInteger.valueOf(i.toLong())
    return result
}

fun nExact(X: Int, Y: Int, Z: Int): BigInteger {
    if ((X and 1) != (Y and 1) || (Y and 1) != (Z and 1)) return BigInteger.ZERO
    val n = X + Y + Z
    val total = bigFactorial(n) / (bigFactorial(X) * bigFactorial(Y) * bigFactorial(Z))
    val diff = qbinomMinus1Int(n, X) * qbinomMinus1Int(n - X, Y)
    val sign = if (((X shr 1) + (Y shr 1) + (Z shr 1) and 1) == 0) BigInteger.ONE else BigInteger.valueOf(-1)
    return (total + sign * diff) / BigInteger.TWO
}

fun modpow(baseVal: Long, exp: Long, mod: Long): Long {
    var result = 1L
    var b = baseVal % mod
    var e = exp
    while (e > 0L) {
        if (e and 1L == 1L) result = (result * b) % mod
        b = (b * b) % mod
        e = e shr 1
    }
    return result
}

fun mainSolve(): Long {
    val assert1 = nExact(2, 2, 2)
    check(assert1 == BigInteger.valueOf(42L)) { "Assert 1 failed: $assert1" }
    val assert2 = nExact(8, 8, 8)
    check(assert2 == BigInteger.valueOf(4732773210L)) { "Assert 2 failed: $assert2" }

    val cubes = IntArray(88) { i -> i * i * i }
    val maxN = 3 * cubes[87]

    val fact = LongArray(maxN + 1)
    fact[0] = 1L
    for (i in 1..maxN) {
        fact[i] = (fact[i - 1] * i.toLong()) % MODVAL
    }

    val invfact = LongArray(maxN + 1)
    invfact[maxN] = modpow(fact[maxN], MODVAL - 2L, MODVAL)
    for (i in maxN downTo 1) {
        invfact[i - 1] = (invfact[i] * i.toLong()) % MODVAL
    }

    val inv2 = (MODVAL + 1L) / 2L

    val halves = IntArray(88) { ai -> cubes[ai] shr 1 }
    val invf = LongArray(88) { ai -> invfact[cubes[ai]] }
    val par = IntArray(88) { i -> i and 1 }

    fun combMod(n: Int, k: Int): Long {
        if (k < 0 || k > n) return 0L
        return fact[n] * invfact[k] % MODVAL * invfact[n - k] % MODVAL
    }

    var totalSum = 0L
    for (ai in 0 until 88) {
        val X = cubes[ai]
        val hx = halves[ai]
        val invX = invf[ai]
        val px = par[ai]
        for (bj in 0 until 88) {
            val Y = cubes[bj]
            val hy = halves[bj]
            val invY = invf[bj]
            val py = par[bj]
            for (ck in 0 until 88) {
                if (px != py || py != par[ck]) continue
                val Z = cubes[ck]
                val hz = halves[ck]
                val invZ = invf[ck]
                val n = X + Y + Z
                var T = fact[n]
                T = (T * invX) % MODVAL
                T = (T * invY) % MODVAL
                T = (T * invZ) % MODVAL
                val D: Long
                if ((n and 1) == 0 && (X and 1) == 1) {
                    D = 0L
                } else {
                    val D1 = combMod(n shr 1, hx)
                    val n2 = n - X
                    if ((n2 and 1) == 0 && (Y and 1) == 1) {
                        D = 0L
                    } else {
                        val D2 = combMod(n2 shr 1, hy)
                        D = (D1 * D2) % MODVAL
                    }
                }
                val Nmod: Long
                if (((hx + hy + hz) and 1) == 0) {
                    Nmod = ((T + D) % MODVAL * inv2) % MODVAL
                } else {
                    Nmod = (((T - D) % MODVAL + MODVAL) % MODVAL * inv2) % MODVAL
                }
                totalSum += Nmod
                if (totalSum >= MODVAL) totalSum -= MODVAL
            }
        }
    }
    return totalSum % MODVAL
}

fun main() {
    println(mainSolve())
}

File: 981/981.cs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/981.py

using System;
using System.Numerics;

class Program981 {
    const long MODVAL = 888888883L;

    static BigInteger QbinomMinus1Int(int n, int k) {
        if ((n & 1) == 0 && (k & 1) == 1) return BigInteger.Zero;
        return BigComb(n >> 1, k >> 1);
    }

    static BigInteger BigComb(int n, int k) {
        if (k < 0 || k > n) return BigInteger.Zero;
        if (k == 0 || k == n) return BigInteger.One;
        BigInteger num = BigInteger.One;
        BigInteger den = BigInteger.One;
        int kk = k < n - k ? k : n - k;
        for (int i = 0; i < kk; i++) {
            num = num * (n - i);
            den = den * (i + 1);
        }
        return num / den;
    }

    static BigInteger BigFactorial(int n) {
        BigInteger result = BigInteger.One;
        for (int i = 2; i <= n; i++) result *= i;
        return result;
    }

    static BigInteger NExact(int X, int Y, int Z) {
        if ((X & 1) != (Y & 1) || (Y & 1) != (Z & 1)) return BigInteger.Zero;
        int n = X + Y + Z;
        BigInteger total = BigFactorial(n) / (BigFactorial(X) * BigFactorial(Y) * BigFactorial(Z));
        BigInteger diff = QbinomMinus1Int(n, X) * QbinomMinus1Int(n - X, Y);
        BigInteger sign = (((X >> 1) + (Y >> 1) + (Z >> 1)) & 1) == 0 ? BigInteger.One : BigInteger.MinusOne;
        return (total + sign * diff) / 2;
    }

    static long Modpow(long baseVal, long exp, long mod) {
        long result = 1L;
        long b = baseVal % mod;
        long e = exp;
        while (e > 0L) {
            if ((e & 1L) == 1L) result = (long)((long)result * b % mod);
            b = (long)((long)b * b % mod);
            e >>= 1;
        }
        return result;
    }

    static long MainSolve() {
        BigInteger assert1 = NExact(2, 2, 2);
        if (assert1 != 42) throw new Exception("Assert 1 failed: " + assert1);
        BigInteger assert2 = NExact(8, 8, 8);
        if (assert2 != 4732773210L) throw new Exception("Assert 2 failed: " + assert2);

        int[] cubes = new int[88];
        for (int i = 0; i < 88; i++) cubes[i] = i * i * i;
        int maxN = 3 * cubes[87];

        long[] fact = new long[maxN + 1];
        fact[0] = 1L;
        for (int i = 1; i <= maxN; i++) {
            fact[i] = (fact[i - 1] * i) % MODVAL;
        }

        long[] invfact = new long[maxN + 1];
        invfact[maxN] = Modpow(fact[maxN], MODVAL - 2L, MODVAL);
        for (int i = maxN; i > 0; i--) {
            invfact[i - 1] = (invfact[i] * i) % MODVAL;
        }

        long inv2 = (MODVAL + 1L) / 2L;

        int[] halves = new int[88];
        for (int ai = 0; ai < 88; ai++) halves[ai] = cubes[ai] >> 1;
        long[] invf = new long[88];
        for (int ai = 0; ai < 88; ai++) invf[ai] = invfact[cubes[ai]];
        int[] par = new int[88];
        for (int i = 0; i < 88; i++) par[i] = i & 1;

        long CombMod(int n, int k) {
            if (k < 0 || k > n) return 0L;
            return fact[n] * invfact[k] % MODVAL * invfact[n - k] % MODVAL;
        }

        long totalSum = 0L;
        for (int ai = 0; ai < 88; ai++) {
            int X = cubes[ai];
            int hx = halves[ai];
            long invX = invf[ai];
            int px = par[ai];
            for (int bj = 0; bj < 88; bj++) {
                int Y = cubes[bj];
                int hy = halves[bj];
                long invY = invf[bj];
                int py = par[bj];
                for (int ck = 0; ck < 88; ck++) {
                    if (px != py || py != par[ck]) continue;
                    int Z = cubes[ck];
                    int hz = halves[ck];
                    long invZ = invf[ck];
                    int n = X + Y + Z;
                    long T = fact[n];
                    T = T * invX % MODVAL;
                    T = T * invY % MODVAL;
                    T = T * invZ % MODVAL;
                    long D;
                    if ((n & 1) == 0 && (X & 1) == 1) {
                        D = 0L;
                    } else {
                        long D1 = CombMod(n >> 1, hx);
                        int n2 = n - X;
                        if ((n2 & 1) == 0 && (Y & 1) == 1) {
                            D = 0L;
                        } else {
                            long D2 = CombMod(n2 >> 1, hy);
                            D = D1 * D2 % MODVAL;
                        }
                    }
                    long Nmod;
                    if (((hx + hy + hz) & 1) == 0) {
                        Nmod = (T + D) % MODVAL * inv2 % MODVAL;
                    } else {
                        Nmod = ((T - D) % MODVAL + MODVAL) % MODVAL * inv2 % MODVAL;
                    }
                    totalSum += Nmod;
                    if (totalSum >= MODVAL) totalSum -= MODVAL;
                }
            }
        }
        return totalSum % MODVAL;
    }

    static void Main(string[] args) {
        Console.WriteLine(MainSolve());
    }
}
